File: backend/app/extractors/net_quantity.py
"""Extract net quantity per Rule 6(1)(c) + Rule 13."""

from __future__ import annotations

import re

from app.domain import ExtractedField, ImageMeta, OCRWord

PATTERN = re.compile(
    r"(\d+(?:\.\d+)?)\s*(g|kg|ml|l|gm|GM|Kg|Litre|Liter|litre|liter|mL|ML)\b", re.I
)
NON_METRIC = re.compile(r"\b(?:oz|ounce|ounces|lb|lbs|pound|pounds|fl\.?\s*oz)\b", re.I)

NET_QTY_ANCHOR = re.compile(
    r"\b(?:net\s*(?:wt\.?|weight|qty\.?|quantity|vol\.?|volume|contents?)|gross\s*(?:wt\.?|weight)|quantity)\b",
    re.I,
)

NUTRITION_EXCLUSION = re.compile(
    r"\b(?:protein|energy|carbohydrate|carbs?|fat|sugar|sodium|cholesterol|nutrition|nutritional|nutrients?|per\s+100\s*g|serve\s+size|serving)\b",
    re.I,
)

BARCODE_PATTERN = re.compile(r"\b\d{8,}\b")

NON_STANDARD_UNIT_PATTERN = re.compile(
    r"\b\d+(?:\.\d+)?\s*(?:gm|gms|g\.|\bkg\.|\bkgs\b|ml\.|lt|ltr)\b",
    re.I,
)


def extract_net_quantity(
    ocr_words: list[OCRWord], image_meta: ImageMeta, allowed_units: list[str]
) -> ExtractedField | None:
    allowed = {u.lower() for u in allowed_units}
    best_candidate: tuple[ExtractedField, int] | None = None  # (field, priority)

    for i in range(len(ocr_words)):
        # Filter out barcode numbers (>= 8 consecutive digits)
        token_text = ocr_words[i].text
        if BARCODE_PATTERN.search(token_text):
            continue

        ws = ocr_words[i : i + 2]
        text = " ".join(w.text for w in ws)
        m = PATTERN.search(text)
        if not (m and m.group(2).lower() in allowed):
            continue

        # Check surrounding context for nutrition table keywords (skip 8g protein, 50g carbs, etc.)
        context_start = max(0, i - 4)
        context_end = min(len(ocr_words), i + 4)
        context_text = " ".join(w.text for w in ocr_words[context_start:context_end])
        is_nutrition = bool(NUTRITION_EXCLUSION.search(context_text))

        # Check if preceded by a statutory Net Quantity anchor keyword
        has_anchor = bool(NET_QTY_ANCHOR.search(context_text))

        candidate = ExtractedField(
            "net_quantity",
            m.group(0),
            ws[0].bbox,
            sum(w.confidence for w in ws) / len(ws),
            [w.bbox for w in ws],
        )

        # Priority 3: Labeled anchor (highest priority)
        # Priority 2: Not in nutrition table
        # Priority 1: In nutrition table (lowest priority fallback)
        priority = 3 if has_anchor else (2 if not is_nutrition else 1)

        if best_candidate is None or priority > best_candidate[1]:
            best_candidate = (candidate, priority)
            if priority == 3:
                break

    if best_candidate is not None:
        return best_candidate[0]

    return ExtractedField("net_quantity", None, None, 0.0, [])

File: frontend/lib/rules/extractors/net-quantity.ts
import type { ExtractedField, ImageMeta, OCRWord } from '../domain';
import { groupWordsIntoSpatialRows } from '../../spatial-layout';
import { avgConfidence } from './base';

export const PATTERN =
  /(\d+(?:\.\d+)?)\s*(g|kg|ml|l|gm|gms|GM|Kg|Litre|Liter|litre|liter|mL|ML|N|U)\b/i;

export const NON_METRIC = /\b(?:oz|ounce|ounces|lb|lbs|pound|pounds|fl\.?\s*oz)\b/i;

const NET_QTY_LABEL = /\b(?:net\s*(?:weight|qty\.?|quantity|contents?|vol\.?|volume|wt\.?))\b/i;

export function extractNetQuantity(
  ocrWords: OCRWord[],
  imageMeta: ImageMeta,
  allowedUnits: string[]
): ExtractedField {
  const allowed = new Set(allowedUnits.map((u) => u.toLowerCase()));
  allowed.add('gm');
  allowed.add('gms');

  // A bare metric value may be a serving size or nutrition value. Only accept a
  // quantity when the same visual row explicitly identifies it as net contents.
  const spatialRows = groupWordsIntoSpatialRows(ocrWords);
  for (const row of spatialRows) {
    if (
      NET_QTY_LABEL.test(row.text) &&
      !/\b(?:serve|serving|per\s*100|nutrition)\b/i.test(row.text)
    ) {
      console.log(`[extractNetQuantity] Checking labeled spatial row: "${row.text}"`);
      const m = PATTERN.exec(row.text);
      if (m && allowed.has(m[2].toLowerCase())) {
        const val = m[0];
        const valueIndex = row.words.findIndex((word) => word.text.includes(m[1]));
        const valWord = row.words[valueIndex] ?? row.words[0];
        const escapedUnit = m[2].replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        const unitWord =
          row.words
            .slice(Math.max(0, valueIndex))
            .find((word) => new RegExp(`^${escapedUnit}\\.?$`, 'i').test(word.text)) ?? valWord;
        const evidenceWords = unitWord === valWord ? [valWord] : [valWord, unitWord];
        console.log(`[extractNetQuantity] Found labeled net quantity: "${val}" in "${row.text}"`);
        return {
          name: 'net_quantity',
          value: val,
          bbox: valWord.bbox,
          confidence: avgConfidence(evidenceWords),
          evidence_spans: evidenceWords.map((word) => word.bbox),
        };
      }
    }
  }

  // OCR sometimes emits the label and value as adjacent blocks rather than one row.
  for (let i = 0; i < ocrWords.length; i++) {
    if (!NET_QTY_LABEL.test(ocrWords[i].text)) continue;
    const ws = ocrWords.slice(i, i + 8);
    const text = ws.map((word) => word.text).join(' ');
    if (/\b(?:serve|serving|per\s*100|nutrition)\b/i.test(text)) continue;
    const match = PATTERN.exec(text);
    if (!match || !allowed.has(match[2].toLowerCase())) continue;
    const valueWord = ws.find((word) => PATTERN.test(word.text)) ?? ws[0];
    return {
      name: 'net_quantity',
      value: match[0],
      bbox: valueWord.bbox,
      confidence: Math.min(...ws.map((word) => word.confidence)),
      evidence_spans: ws.map((word) => word.bbox),
    };
  }

  console.log('[extractNetQuantity] No Net Quantity found in evidence.');
  return {
    name: 'net_quantity',
    value: null,
    bbox: null,
    confidence: 0.0,
    evidence_spans: [],
  };
}
